### ISSUE TITLE
Lifter fails to propagate values for partial registers

### ISSUE THREAD

**Issue by fluxchief** [Vector35/binaryninja-api#921]
When using subregisters (e.g. `ax` vs `eax` vs `rax` or `R30` and `R31` vs `Z`) constant value propagation does not seem to work as intended (even if querying the subregister):

    >>> j
    <il: r31 = r25>
    >>> j.get_reg_value('r25')
    <const 0x6>
    >>> j.get_reg_value_after('r31')
    <undetermined>
    >>> j.get_possible_reg_values_after('r31')
    <const 0x6>
    # Query the 'parent' register:
    >>> j.get_reg_value_after('Z')
    <undetermined>
    >>> j.get_possible_reg_values_after('Z')
    <const 0x600>

**Comment by bpotchik** [comment]
This should be resolved with the latest builds on dev. Can you verify the correct behavior for your test case and let us know?  Thanks!

**Comment by fluxchief** [comment]
Hovering over the combined register (Z) does show the correct value. However it still seems like it does not propagate correctly:
![nolift](https://user-images.githubusercontent.com/2988703/36550452-a222ce48-17f5-11e8-8bdd-a9b01df7cf90.png)
Both values are constant, so I would have expected a reference to [0x100050] instead of showing it as [0x100000 + 0x50]

I played a little with the IL and it seems like the register is set to undefined? somehow in the IL (even if the UI shows otherwise):

```
>>> instructions
[<il: r30 = r24>, <il: r31 = r25>, <il: [0x100000 + Z].b = r18>]
>>> instructions[2].get_reg_value('r24')
<const 0x50>
>>> instructions[2].get_reg_value('r25')
<const 0x0>
>>> instructions[2].get_reg_value('Z')
<undetermined>
>>> instructions[2].get_reg_value('r30')
<undetermined>
>>> instructions[2].get_reg_value('r31')
<undetermined>
```

Which seems to be caused by:
>>> instructions[0]
<il: r30 = r24>
>>> instructions[0].get_reg_value('r24')
<const 0x50>
>>> instructions[0].get_reg_value_after('r30')
<undetermined>

**Comment by bpotchik** [comment]
You should get the correct value with the get_possible_reg_values API as this is what hover uses. If r31, or z was fully defined as a constant then you could use get_reg_value_after to get the value of r30. This fast path(get_reg_) API requires the entire register to be defined in order for constant propagation to work.

**Comment by fluxchief** [comment]
This should be the case?
```
>>> instructions[0]
<il: r30 = r24>
>>> instructions[0].get_reg_value('r24')
<const 0x50>
>>> instructions[0].get_reg_value_after('r30')
<undetermined>
```

r24 is set to const  0x50, r30 is undefined afterwards which doesn't sound right - or am I understanding sth wrong?

**Comment by bpotchik** [comment]
Yes that is correct if r31 is undetermined or not a constant. If I'm not mistaken, r30 and r31 are both partial registers that make up register Z.

If you call get_possible_reg_values the correct value will be returned. Also if you set r31 to a constant then `instructions[0].get_reg_value_after('r30')` should not return undetermined.

**Comment by fluxchief** [comment]
You're right about `get_possible_reg_values` returning the correct value. So why does it make a difference that it is part of a bigger register? The partial register content should be known anyway, even when it's part of a bigger reg.

> Also if you set r31 to a constant then instructions[0].get_reg_value_after('r30') should not return undetermined.
```
>>> instructions[1].get_reg_value_after('r30')
<undetermined>
```
That's sadly not the case (r30 is set in 0, r31 is set in 1)

Just to summarize my test case (maybe I missed some critical information):
Z (2 bytes) consists out of r31 and r30.
Initial reg values are r24 = 0x50, r25 = 0x00.
Three IL instructions:
0: `r30 = r24` (= 0x50, first part of Z const)
1: `r31 = r25` (= 0x00, second part of Z const)
2: ` [0x100000 + Z].b = r18` (= 0x1000000 + 0x0050, *fully const*)
This means after 1, Z should be fully defined as const. But it isn't (in the fast path, the get_possible_reg_values returns the right value).

So, after 0, the content r30 should be known in the fast path as it is fully defined, but it's only available in the get_possible_reg_values.
After 1, the same is true for r31. Now, Z should also be fully defined and available in the fast path, but it's not.

**Comment by bpotchik** [comment]
Hmm, regarding "Also if you set r31 to a constant" could you try to set Z, then set either r30 or r31?

**Comment by fluxchief** [comment]
Setting Z first and then changing the subregs does work - but this is meaningless for me.
There is no way of setting (X/Y/)Z directly in AVR - this is only possible by accessing the subregisters.

**Comment by bpotchik** [comment]
Setting Z was just an exercise to verify that constant propagation works as expected for the get_reg_value API. You will need to use the get_possible_reg_values API in this case.

**Comment by fluxchief** [comment]
Are there plans to change this in the future? Cause this is a blocker for me using subregisters. There is no gain as long as it decreases the autoanalysis capabilities.

**Comment by bpotchik** [comment]
Why is this a blocker? What specifically regarding autoanalysis capabilities is broken? Perhaps there is a different bug regarding analysis with regards to falling back to the slower dataflow API.

**Comment by fluxchief** [comment]
When I use subregisters, the Z register is not fully defined (because of the optimization you've described). When I don't use them but generate this register by myself by using shifts and or's, the value is calculated properly and the correct reference is shown.

With using subregisters:
![1](https://user-images.githubusercontent.com/2988703/36718227-51094e74-1ba1-11e8-9b22-ecadd117d18d.png)
![2](https://user-images.githubusercontent.com/2988703/36718228-512033fa-1ba1-11e8-9ddf-723284403492.png)

Without:
![3](https://user-images.githubusercontent.com/2988703/36718226-50ee7996-1ba1-11e8-879b-795705f1b4ec.png)
![4](https://user-images.githubusercontent.com/2988703/36718225-50d8a378-1ba1-11e8-81d1-c88daa431ded.png)

The OSC_CTRL reference is lost when making use of the subregs.

**Comment by toolCHAINZ** [comment]
This isn't exactly the same issue, but it was close enough that I felt bad opening a new issue for it.

I have some code making a jumptable where the lowest byte of the register holding the jump table index is compared to a constant value when deciding whether to take the jumptable. If the value is less than the constant, then the low byte is moved and zero extended to another register which is subsequently used to calculate the jump address.

If I (incorrectly) lift the comparison instruction to use `il.reg` without any masking, the value of the register holding the index is successfully propagated. If, however, I use `il.lowpart` or `il.and` to isolate part of the register for the comparison, the value of that same portion of the register is not constrained when accessed for calculating the jump address.

This leads to the dataflow thinking that, since the index is unconstrained, there are 255 possible jump targets instead of 5 (in the case I am seeing).

**Comment by fluxchief** [comment]
Is there any update on this? I'd really like to start using register pairs to make the code more readable.

**Comment by bpotchik** [comment]
Actually we talked about this one today. This definitely won't land until after 1.2 is out. However we believe there is a path to add support for this.

**Comment by jrozner** [comment]
1.2 has been out for a while. Has there been any updates toward addressing this?

**Comment by bpotchik** [comment]
Same issue here. #2766

**Comment by bpotchik** [comment]
Fixed in `3.2.3934-dev`.

I did not mark constant propagated partial reg expressions for elimination yet, but will probably push that shortly.
